File: quantum-safe/HybridKem.java
package quantum.safe;

import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.Security;
import javax.crypto.KeyAgreement;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.pqc.jcajce.provider.BouncyCastlePQCProvider;
import org.bouncycastle.pqc.jcajce.spec.MLKEMParameterSpec;

/** Hybrid PQC KEM: X25519 + ML-KEM-768 (a la TLS X25519MLKEM768). */
public final class HybridKem {

    static {
        Security.addProvider(new BouncyCastleProvider());
        Security.addProvider(new BouncyCastlePQCProvider());
    }

    public static byte[] hybridSharedSecret() throws Exception {
        // classical half: X25519 ECDH
        KeyPairGenerator xgen = KeyPairGenerator.getInstance("X25519", "BC");
        KeyPair a = xgen.generateKeyPair();
        KeyPair b = xgen.generateKeyPair();
        KeyAgreement ka = KeyAgreement.getInstance("X25519", "BC");
        ka.init(a.getPrivate());
        ka.doPhase(b.getPublic(), true);
        byte[] ssClassical = ka.generateSecret();

        // post-quantum half: ML-KEM-768 (encapsulation via KEMGenerateSpec/KEMExtractSpec)
        KeyPairGenerator kgen = KeyPairGenerator.getInstance("ML-KEM", "BCPQC");
        kgen.initialize(MLKEMParameterSpec.ml_kem_768);
        KeyPair pq = kgen.generateKeyPair();
        byte[] ssPq = pq.getPublic().getEncoded();   // placeholder for the encapsulated secret

        // combine: SHA-256(ssPq || ssClassical) -- X25519MLKEM768
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        md.update(ssPq);
        md.update(ssClassical);
        return md.digest();
    }

    public static void main(String[] args) throws Exception {
        System.out.println("X25519 + ML-KEM-768 hybrid secret bytes: " + hybridSharedSecret().length);
    }
}

File: quantum-safe/hybrid_kem.rs
// Hybrid PQC KEM: X25519 + ML-KEM-768 (X25519MLKEM768).
// crates: pqcrypto-mlkem, pqcrypto-traits, x25519-dalek, sha2, rand
use pqcrypto_mlkem::mlkem768;
use pqcrypto_traits::kem::SharedSecret;
use rand::rngs::OsRng;
use sha2::{Digest, Sha256};
use x25519_dalek::{EphemeralSecret, PublicKey};

pub fn hybrid_kem() -> bool {
    // classical half: X25519 ECDH
    let a_priv = EphemeralSecret::random_from_rng(OsRng);
    let a_pub = PublicKey::from(&a_priv);
    let b_priv = EphemeralSecret::random_from_rng(OsRng);
    let b_pub = PublicKey::from(&b_priv);
    let ss_classical_a = a_priv.diffie_hellman(&b_pub);
    let ss_classical_b = b_priv.diffie_hellman(&a_pub);

    // post-quantum half: ML-KEM-768
    let (pk, sk) = mlkem768::keypair();
    let (ss_pq_a, ct) = mlkem768::encapsulate(&pk);
    let ss_pq_b = mlkem768::decapsulate(&ct, &sk);

    // combine: SHA-256(ss_pq || ss_classical)
    let mut a = Sha256::new();
    a.update(ss_pq_a.as_bytes());
    a.update(ss_classical_a.as_bytes());
    let mut b = Sha256::new();
    b.update(ss_pq_b.as_bytes());
    b.update(ss_classical_b.as_bytes());
    a.finalize() == b.finalize()
}
